consensus/propagation/have_wants.go

```diff
@@ -63,58 +66,143 @@ func (blockProp *Reactor) handleHaves(peer p2p.ID, haves *proptypes.HaveParts) {
+	// TODO implement
+	return 300
+}
+
+func (blockProp *Reactor) wantsSendingRoutine(ps *PeerState) {
+	var requestsBA *bits.BitArray
+	height := int64(0)
+	round := int32(0)
+	periodicRequestsDelay := 10 * time.Millisecond
+	periodicRequestTicker := time.NewTicker(periodicRequestsDelay)
+	for {
+		select {
+		case <-blockProp.ctx.Done():
+			return
+		case <-periodicRequestTicker.C:
+			// periodically send the existing requests
+			if requestsBA != nil && !requestsBA.IsEmpty() {
+				blockProp.sendWant(ps, height, round, requestsBA)
+				requestsBA = nil
+			}
+		case req, ok := <-ps.requestChan:
+			if !ok {
+				return
+			}
+
+			// concurrent per-peer per-block request limit reached
+			if ps.requestCount.Load() >= perPeerPerBlockConcurrentRequestLimit() {
+				// we reached the limit of requests, we request what we can
+				if requestsBA != nil && !requestsBA.IsEmpty() {
+					blockProp.sendWant(ps, height, round, requestsBA)
+					requestsBA = nil
+				}
+				// wait for a part to be received before continuing
+				select {
+				case <-blockProp.ctx.Done():
+					return
+				case _, ok = <-ps.receivedPart:
+					if !ok {
+						return
+					}
+				}
+			}
+
+			// if we're at a new height, we can drop the previous requests
+			if height != req.height || (height == req.height && round != req.round) {
+				requestsBA = nil
+				height = req.height
+				round = req.round
+			}
+
+			partIndex := req.index
+			_, parts, fullReqs, has := blockProp.getAllState(height, round, false)
+			if !has {
+				blockProp.Logger.Error("couldn't find proposal when filtering requests", "height", height, "round", round)
+				continue
+			}
+			if parts.BitArray().GetIndex(int(partIndex)) {
+				continue
+			}
+
+			reqLimit := 1
+			reqs := blockProp.countRequests(height, round, int(partIndex))
+			if len(reqs) >= reqLimit {
+				fullReqs.SetIndex(int(partIndex), true)
+				continue
+			}
+
+			// don't request the part from this peer if we've already requested it
+			// from them.
+			for _, p := range reqs {
+				// p == peer means we have already requested the part from this peer.
+				if p == ps.peer.ID() {
+					continue
+				}
+			}
+
+			if requestsBA == nil {
+				requestsBA = bits.NewBitArray(int(parts.Total()))
+			}
+			requestsBA.SetIndex(int(partIndex), true)
+
+			if len(requestsBA.GetTrueIndices()) >= wantBatchSize(int(parts.Total())) {
+				// no need to keep holding to a large number of requests
+				blockProp.sendWant(ps, height, round, requestsBA)
+				requestsBA = nil
 			}
 		}
 	}
+}
 
-	if hc.IsEmpty() {
-		return
-	}
+// wantBatchSize returns the maximum number of parts to request in a batch.
+// this ensures sending requests without waiting too long especially for small blocks.
+func wantBatchSize(partsCount int) int {
+	return partsCount * 5 / 100
+}
 
-	// send a want back to the sender of the haves with the wants we
+func (blockProp *Reactor) sendWant(ps *PeerState, height int64, round int32, requestsBA *bits.BitArray) {
```

Comment:
[naming opinion/tip]

the type is right next to the var name, BA just takes up space. in general, I don't think naming things after types is good unless its a python script or something. `wants`, `reqs`, or even just `r` is much easier to read imo

Reply:
9721c137d31713a8e50acea83b9517f2e20378c9